src/client/pythonEnvironments/common/environmentIdentifier.ts

```diff
@@ -122,6 +113,114 @@ async function isWindowsStoreEnvironment(interpreterPath: string): Promise<boole
     return false;
 }
 
+/**
+ * Returns the path to Pipfile associated with the provided directory.
+ * @param cwd the directory to look into
+ * @param lookIntoParentDirectories set to true if we should also search for Pipfile in parent directory
+ */
+async function getAssociatedPipfile(cwd: string, lookIntoParentDirectories: boolean): Promise<string | undefined> {
+    const pipFileName = getEnvironmentVariable('PIPENV_PIPFILE') || 'Pipfile';
+    let depthToSearch = 1;
+    if (lookIntoParentDirectories) {
+        // PIPENV_MAX_DEPTH tells pipenv the maximum number of directories to recursively search for
+        // a Pipfile, defaults to 3: https://pipenv.pypa.io/en/latest/advanced/#pipenv.environments.PIPENV_MAX_DEPTH
+        const maxDepth = getEnvironmentVariable('PIPENV_MAX_DEPTH');
+        if (maxDepth) {
+            depthToSearch = +maxDepth;
+        } else {
+            depthToSearch = 3;
+        }
+    }
+    while (depthToSearch > 0 && cwd !== path.dirname(cwd)) {
+        const pipFile = path.join(cwd, pipFileName);
+        // eslint-disable-next-line no-await-in-loop
+        if (await pathExists(pipFile)) {
+            return pipFile;
+        }
+        cwd = path.dirname(cwd);
+        depthToSearch -= 1;
+    }
+    return undefined;
+}
+
+/**
+ * Returns true if interpreter path belongs to a pipenv environment which is located inside a project, false otherwise.
+ * @param interpreterPath Absolute path to any python interpreter.
+ */
+async function isLocalPipenvEnvironment(interpreterPath: string): Promise<boolean> {
+    // Local pipenv environments are created by setting PIPENV_VENV_IN_PROJECT to 1, which always names the environment
+    // folder '.venv': https://pipenv.pypa.io/en/latest/advanced/#pipenv.environments.PIPENV_VENV_IN_PROJECT
+    // This is the layout we wish to verify.
+    // project
+    // |__ Pipfile  <--- check if Pipfile exists here
+    // |__ .venv    <--- check if name of the folder is '.venv'
+    //     |__ Scripts/bin
+    //         |__ python  <--- interpreterPath
+    const venvFolderName = path.basename(path.dirname(path.dirname(interpreterPath)));
+    if (venvFolderName !== '.venv') {
+        return false;
+    }
+    const directoryWhereVenvResides = path.dirname(path.dirname(path.dirname(interpreterPath)));
```

Comment:
```suggestion
    const directoryWhereVenvResides = path.dirname(venvFolder);
```
This simplifies the `dirname` chain which is hard to follow.